### ISSUE TITLE
Support for Multi-Channel Imagery

### ISSUE THREAD

**Issue by Supersak80** [rbgirshick/py-faster-rcnn#232]
Hi all,
Does anyone know how easy it would be to modify the data layer portion of the code to accept images with more than 3 channels? I'm doing some experiments where I have imagery that is composed of 4 or 5 channels.

**Comment by GBJim** [comment]
Hi @Supersak80: 
I haven't tried it, but I guess what you need is to modify the network definition in training.prototxt  and testing.prototxt.  Then you would train the whole network from ImageNnet and PASCAL, because the given pre-trained weights are for 3 channels only.  Maybe there is a way to transfer the weights from 3 channels into higher channels, I am now sure about it.

**Comment by Supersak80** [comment]
@GBJim I agree with you, the training and testing prototxts need to be modified, but beyond that, the Python data layer code needs to be modified in several places because it's hard coded to allocate tensors for 3 channels only.

**Comment by GBJim** [comment]
@Supersak80, So modification is way more complicated than I thought. 
I am not sure if this is a good idea in your scenario, but is it possible to reduce your higher dimensional channels into 3 dimensions by pre-processing?
Maybe techniques such like PCA, Random Project or some Computer Vision specific methods.

**Comment by Supersak80** [comment]
@GBJim You could probably invoke a dimensionality reduction method, but then it defeats the purpose of utilizing all of the original pixel information to perform an "early fusion" through the CNN. Plus, it's not clear that applying a dimensionality reduction algorithm will preserve the 2D spatial structure of the imagery...

**Comment by vgenty** [comment]
I've been able to train with up to a 12 channel image without too much difficulty. It's slightly more difficult than just changing your prototxt. As @Supersak80 mentions you will have to modify the python data layers. I don't think this should be an open "issue", it's up to the user to modify the framework for their application and I think it's obvious enough where to modify things.

**Comment by Supersak80** [comment]
@vgenty Hi Victor, were you able to generalize the appropriate pieces of code in the python data layer to accept an arbitrary number of channels beyond 3? If so, if you could please give me some pointers and/or code tips on how to do this, it would be greatly appreciated.

**Comment by zhenni** [comment]
@vgenty 
Do you train from sketch? Do you try to initialize the weights with some existing model? If you train with initialization from some existing model, do you make some modification with the  weights of the model??

@Supersak80 
The mainly code to change is in `lib/roi_data_layer/minibatch.py` for training and `lib/fast_rcnn/test.py` if you also want to test.

**Comment by DongXingshuai** [comment]
@vgenty Hello, I wonder if you can post the codes for training multi-channel images or point out which place I shoud change? Many thanks!

**Comment by DongXingshuai** [comment]
@Supersak80 Hello, have you solve this problem?
